frontend/src/components/editor/Cell.tsx

```diff
@@ -311,10 +313,82 @@ const CellComponent = (
     [cellRef, editorView],
   );
 
+  const isCellCodeShown = !cellConfig.hide_code || temporarilyVisible;
+  const isMarkdown =
+    getCurrentLanguageAdapter(editorView.current) === "markdown";
+  const isMarkdownCodeHidden = isMarkdown && !isCellCodeShown;
+
+  const cellContainerRef = useRef<HTMLDivElement>(null);
+
+  // DOM node where the editorView will be mounted
+  const editorViewParentRef = useRef<HTMLDivElement>(null);
+
+  // if the cell is hidden, show the code editor temporarily
+  const temporarilyShowCode = useCallback(async () => {
+    if (!isCellCodeShown && setTemporarilyVisible !== undefined) {
+      setTemporarilyVisible(true);
+      editorView.current?.focus();
+      // Reach one parent up
+      const parent = editorViewParentRef.current?.parentElement;
+      const abortController = new AbortController();
+      parent?.addEventListener(
+        "focusout",
+        () => {
+          requestAnimationFrame(() => {
+            // Skip closing if the focus is still in the parent element
+            const focusedElement = document.activeElement;
+            if (parent?.contains(focusedElement)) {
+              return;
+            }
+            // Hide the code editor
+            setTemporarilyVisible(false);
+            editorView.current?.dom.blur();
+            abortController.abort();
+          });
+        },
+        { signal: abortController.signal },
+      );
+    }
+  }, [isCellCodeShown, setTemporarilyVisible, editorViewParentRef, editorView]);
+
+  const showHiddenMarkdownCode = () => {
+    if (isMarkdownCodeHidden) {
+      temporarilyShowCode();
+    }
+  };
+
+  // If the cell is too short, we need to position the cellStatus inline to not cause overlaps
+  // This can only happen to markdown cells when the code is hidden completely
+  const [isCellStatusInline, setIsCellStatusInline] = useState(false);
+  useEffect(() => {
+    const isCurrentlyMarkdown =
+      getCurrentLanguageAdapter(editorView.current) === "markdown";
+    if (!isCurrentlyMarkdown) {
+      return;
+    }
+
+    const resizeObserver = new ResizeObserver(() => {
+      const height = cellContainerRef.current?.clientHeight;
+      if (height && height < 60) {
+        setIsCellStatusInline(true);
+      } else {
+        setIsCellStatusInline(false);
+      }
+    });
+
+    if (cellContainerRef.current) {
+      resizeObserver.observe(cellContainerRef.current);
+    }
+
+    return () => {
+      resizeObserver.disconnect();
+    };
+  }, [isMarkdown]);
```

Comment:
this file is getting long (my fault), are there any utils you can extract out for readability (and testability), e.g.:

```ts
import { useEffect, RefObject } from 'react';

interface UseResizeObserverOptions {
  ref: RefObject<Element>;
  skip?: boolean;
  onResize: (entry: ResizeObserverEntry) => void;
}

export function useResizeObserver({
  ref,
  skip = false,
  onResize
}: UseResizeObserverOptions) {
  useEffect(() => {
    if (skip || !ref.current) return;

    const resizeObserver = new ResizeObserver((entries) => {
      const entry = entries[0];
      if (entry) {
        onResize(entry);
      }
    });

    resizeObserver.observe(ref.current);

    return () => {
      resizeObserver.disconnect();
    };
  }, [ref, skip, onResize]);
};
```